crates/services/src/seqlock.rs

```diff
@@ -0,0 +1,126 @@
+//! A simple implementation of a sequential lock.
+//! More details: <https://docs.kernel.org/locking/seqlock.html>
+//! Optimized for occasional writes and frequent reads
+
+use std::{
+    cell::UnsafeCell,
+    sync::atomic::{
+        AtomicU64,
+        Ordering,
+    },
+};
+
+/// A simple implementation of a sequential lock.
+/// some usage of unsafe, T must be Copy
+#[derive(Debug)]
+pub struct SeqLock<T> {
+    sequence: AtomicU64,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send> Sync for SeqLock<T> {}
+
+impl<T> SeqLock<T> {
+    /// Create a new SeqLock with the given data
+    pub fn new(data: T) -> Self {
+        Self {
+            sequence: AtomicU64::new(0),
+            data: UnsafeCell::new(data),
+        }
+    }
+
+    /// Write the data
+    pub fn write<F>(&self, f: F)
+    where
+        F: FnOnce(&mut T),
+    {
+        // starting guard
+        self.sequence.fetch_add(1, Ordering::Release);
+
+        // Modify the data
+        unsafe {
+            f(&mut *self.data.get());
+        }
+
+        // ending guard
+        self.sequence.fetch_add(1, Ordering::Release);
+    }
+
+    /// Read the data
+    pub fn read(&self) -> T
+    where
+        T: Copy,
+    {
+        loop {
+            // check starting guard
+            let start = self.sequence.load(Ordering::Acquire);
+
+            // if odd, write in progress
+            if start % 2 != 0 {
+                continue;
+            }
+
+            let data = unsafe { *self.data.get() };
+
+            // check starting/ending guard
+            let end = self.sequence.load(Ordering::Acquire);
+
+            // if value changed, retry
+            if start == end && start % 2 == 0 {
+                return data;
+            }
```

Comment:
nit: synchronization is required only when the start and end guard coincide. See https://doc.rust-lang.org/std/sync/atomic/fn.fence.html#examples for an example.

(But I suspect we don't need to optimize further re your implementation, so happy to leave it as is)

```suggestion
            let end = self.sequence.load(Ordering::Relaxed);

            // if value changed, retry
            if start == end && start % 2 == 0 {
                std::sync::atomic::fence(Ordering::Acquire)
                return data;
            }
```

Reply:
yep, it's probably an overkill :)